System Channels in Channel Groups

When using [System-Channels](http://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#channels) within channel groups, for only one Thing it is not possible to bind any things for the whole binding.

The PaperUI shows any thing of this binding as "Unknown Decvice" and reports "409 Conflict" when trying to add the device as thing. Within the logfile of OpenHab you can find the following messages:

```
2016-06-20 09:16:46.146 [ERROR] [s.c.thing.binding.ThingFactory:112  ] - Thing factory (class MyThingHandlerFactory) returned null on create thing when it reports to support the thing type (binding:myThing).
2016-06-20 09:16:46.146 [WARN ] [s.c.d.internal.PersistentInbox:152  ] - Cannot create thing. No binding found that supports creating a thing of type binding:myThing.
```

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/internal/XmlHelperTest.groovy
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/internal/XmlHelperTest.groovy
@@ -1,0 +1,31 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.thing.xml.internal
+
+import static org.hamcrest.CoreMatchers.*
+import static org.junit.Assert.*
+import static org.junit.matchers.JUnitMatchers.*
+
+import org.junit.Test
+
+/**
+ * @author Simon Kaufmann - Initial contribution and API
+ *
+ */
+class XmlHelperTest {
+
+    @Test
+    void 'assert that dot is converted to colon'() {
+        assertThat XmlHelper.getSystemUID("system.test"), is("system:test")
+    }
+
+    @Test
+    void 'assert that system namespace is prepended if no prefix was given'() {
+        assertThat XmlHelper.getSystemUID("test"), is("system:test")
+    }
+}

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/test/SystemChannelsInChannelGroupsTest.groovy
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/test/SystemChannelsInChannelGroupsTest.groovy
@@ -1,0 +1,114 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.thing.xml.test
+
+import static org.hamcrest.CoreMatchers.*
+import static org.junit.Assert.*
+import static org.junit.matchers.JUnitMatchers.*
+
+import org.eclipse.smarthome.core.thing.binding.ThingTypeProvider
+import org.eclipse.smarthome.core.thing.type.ChannelGroupTypeUID
+import org.eclipse.smarthome.core.thing.type.ChannelTypeRegistry
+import org.eclipse.smarthome.core.thing.type.ThingType
+import org.eclipse.smarthome.test.OSGiTest
+import org.eclipse.smarthome.test.SyntheticBundleInstaller
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.osgi.framework.Bundle
+
+
+/**
+ * @author Simon Kaufmann - Initial contribution and API
+ *
+ */
+class SystemChannelsInChannelGroupsTest extends OSGiTest {
+
+    static final String SYSTEM_CHANNELS_IN_CHANNEL_GROUPS_BUNDLE_NAME = "SystemChannelsInChannelGroups.bundle"
+
+    ThingTypeProvider thingTypeProvider
+    ChannelTypeRegistry channelTypeRegistry
+
+
+    @Before
+    void setUp() {
+        thingTypeProvider = getService(ThingTypeProvider)
+        assertThat thingTypeProvider, is(notNullValue())
+
+        channelTypeRegistry = getService(ChannelTypeRegistry)
+        assertThat channelTypeRegistry, is(notNullValue())
+    }
+
+    @After
+    void tearDown() {
+        SyntheticBundleInstaller.uninstall(getBundleContext(), SYSTEM_CHANNELS_IN_CHANNEL_GROUPS_BUNDLE_NAME)
+    }
+
+    @Test
+    void 'assert that System Channels used within channel groups are loaded and unloaded'() {
+        def bundleContext = getBundleContext()
+        def initialNumberOfThingTypes = thingTypeProvider.getThingTypes(null).size()
+        def initialNumberOfChannelTypes = channelTypeRegistry.getChannelTypes().size()
+        def initialNumberOfChannelGroupTypes = channelTypeRegistry.getChannelGroupTypes().size()
+
+        // install test bundle
+        Bundle bundle = SyntheticBundleInstaller.install(bundleContext, SYSTEM_CHANNELS_IN_CHANNEL_GROUPS_BUNDLE_NAME)
+        assertThat bundle, is(notNullValue())
+
+        def thingTypes = thingTypeProvider.getThingTypes(null)
+        assertThat thingTypes.size(), is(initialNumberOfThingTypes + 1)
+        assertThat channelTypeRegistry.getChannelTypes().size(), is(initialNumberOfChannelTypes + 1)
+        assertThat channelTypeRegistry.getChannelGroupTypes().size(), is(initialNumberOfChannelGroupTypes + 1)
+
+        // uninstall test bundle
+        bundle.uninstall();
+        assertThat bundle.state, is(Bundle.UNINSTALLED)
+
+        assertThat thingTypeProvider.getThingTypes(null).size(), is(initialNumberOfThingTypes)
+        assertThat channelTypeRegistry.getChannelTypes().size(), is(initialNumberOfChannelTypes)
+        assertThat channelTypeRegistry.getChannelGroupTypes().size(), is(initialNumberOfChannelGroupTypes)
+
+    }
+
+    @Test
+    void 'assert that Thing Types with system channels in channel groups have proper channel definitions'() {
+        def bundleContext = getBundleContext()
+
+        // install test bundle
+        Bundle sysBundle = SyntheticBundleInstaller.install(bundleContext, SYSTEM_CHANNELS_IN_CHANNEL_GROUPS_BUNDLE_NAME)
+        assertThat sysBundle, is(notNullValue())
+
+        List<ThingType> thingTypes = thingTypeProvider.getThingTypes(null).findAll{it.getUID().getId().equals("wireless-router") }
+        assertThat thingTypes.size(), is(1)
+
+        def channelGroupTypes = channelTypeRegistry.getChannelGroupTypes()
+
+        def channelGroup = channelGroupTypes.find{it.getUID().equals(new ChannelGroupTypeUID("SystemChannelsInChannelGroups:channelGroup"))}
+        assertThat channelGroup, is(notNullValue())
+
+        def channelDefs = channelGroup.getChannelDefinitions()
+
+        def myChannel = channelDefs.findAll {
+            it.id.equals("test") &&
+                    it.getChannelTypeUID().getAsString().equals("system:my-channel") }
+
+        def sigStr = channelDefs.findAll {
+            it.id.equals("sigstr") &&
+                    it.getChannelTypeUID().getAsString().equals("system:signal-strength") }
+
+        def lowBat = channelDefs.findAll {
+            it.id.equals("lowbat") &&
+                    it.getChannelTypeUID().getAsString().equals("system:low-battery") }
+
+
+        assertThat myChannel.size(), is(1)
+        assertThat sigStr.size(), is(1)
+        assertThat lowBat.size(), is(1)
+    }
+
+}

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/SystemChannelsInChannelGroups.bundle/ESH-INF/binding/binding.xml
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/SystemChannelsInChannelGroups.bundle/ESH-INF/binding/binding.xml
@@ -1,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<binding:binding xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:binding="http://eclipse.org/smarthome/schemas/binding/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 org.eclipse.smarthome.binding.xsd"
+	id="SystemChannels">
+
+	<name>SystemChannelsInChannelGroups Test Binding</name>
+	<description>Used to test System Wide Channels which are used within Channel Groups. Provides definitions of system channels.</description>
+	<author>Deutsche Telekom AG</author>
+	
+</binding:binding>

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/SystemChannelsInChannelGroups.bundle/ESH-INF/thing/thing-types.xml
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/SystemChannelsInChannelGroups.bundle/ESH-INF/thing/thing-types.xml
@@ -1,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="SystemChannelsInChannelGroups" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<!-- YahooWeather Binding -->
+	<thing-type id="wireless-router">
+		<label>Simple wireless router</label>
+		<description>Simple wireless router</description>
+
+		<channel-groups>
+			<channel-group typeId="channelGroup" id="group1" />
+			<channel-group typeId="channelGroup" id="group2" />
+		</channel-groups>
+
+	</thing-type>
+
+	<!-- System Channel Type -->
+	<channel-type id="my-channel" system="true">
+		<item-type>Number</item-type>
+		<label>My Channel</label>
+		<category>Other</category>
+		<tags>
+			<tag>
+				Test
+			</tag>
+		</tags>
+		<config-description>
+			<parameter name="range" type="integer" readOnly="true" min="0" max="4" step="1" />
+		</config-description>
+	</channel-type>
+
+	<!-- Channel Group -->
+	<channel-group-type id="channelGroup">
+		<label>Test Channel Group Label</label>
+		<description>Test Channel Group Description</description>
+		<channels>
+			<channel id="sigstr" typeId="system.signal-strength" />
+			<channel id="test" typeId="system.my-channel" />
+			<channel id="lowbat" typeId="system.low-battery" />
+		</channels>
+	</channel-group-type>
+
+</thing:thing-descriptions>

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/SystemChannelsInChannelGroups.bundle/META-INF/MANIFEST.MF
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/SystemChannelsInChannelGroups.bundle/META-INF/MANIFEST.MF
@@ -1,0 +1,10 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Synthetic Test Bundle
+Bundle-SymbolicName: SystemChannelsInChannelGroups.bundle;singleton:=true
+Bundle-Vendor: Eclipse.org/SmartHome
+Bundle-Version: 1.0.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .
+Import-Package: com.google.common.collect,
+ org.osgi.framework

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelGroupTypeXmlResult.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelGroupTypeXmlResult.java
@@ -14,7 +14,6 @@ import java.util.Map;
 import org.eclipse.smarthome.core.thing.type.ChannelDefinition;
 import org.eclipse.smarthome.core.thing.type.ChannelGroupType;
 import org.eclipse.smarthome.core.thing.type.ChannelGroupTypeUID;
-import org.eclipse.smarthome.core.thing.type.ChannelType;
 
 import com.thoughtworks.xstream.converters.ConversionException;
 
@@ -49,41 +48,26 @@ public class ChannelGroupTypeXmlResult {
         return this.channelGroupTypeUID;
     }
 
-    protected List<ChannelDefinition> toChannelDefinitions(List<ChannelXmlResult> channelTypeReferences,
-            Map<String, ChannelType> channelTypes) throws ConversionException {
+    protected List<ChannelDefinition> toChannelDefinitions(List<ChannelXmlResult> channelTypeReferences)
+            throws ConversionException {
 
         List<ChannelDefinition> channelTypeDefinitions = null;
 
         if ((channelTypeReferences != null) && (channelTypeReferences.size() > 0)) {
             channelTypeDefinitions = new ArrayList<>(channelTypeReferences.size());
 
-            if (channelTypes != null) {
-                for (ChannelXmlResult channelTypeReference : channelTypeReferences) {
-                    String id = channelTypeReference.getId();
-                    String typeId = channelTypeReference.getTypeId();
-
-                    String typeUID = String.format("%s:%s", this.channelGroupTypeUID.getBindingId(), typeId);
-
-                    ChannelType channelType = channelTypes.get(typeUID);
-                    if (channelType != null) {
-                        ChannelDefinition channelDefinition = new ChannelDefinition(id, channelType.getUID());
-                        channelTypeDefinitions.add(channelDefinition);
-                    } else {
-                        throw new ConversionException("The channel type for '" + typeUID + "' is missing!");
-                    }
-                }
-            } else {
-                throw new ConversionException("Missing the definition of channel types!");
+            for (ChannelXmlResult channelTypeReference : channelTypeReferences) {
+                channelTypeDefinitions
+                        .add(channelTypeReference.toChannelDefinition(this.channelGroupTypeUID.getBindingId()));
             }
         }
 
         return channelTypeDefinitions;
     }
 
-    public ChannelGroupType toChannelGroupType(Map<String, ChannelType> channelTypes) throws ConversionException {
-
+    public ChannelGroupType toChannelGroupType() throws ConversionException {
         ChannelGroupType channelGroupType = new ChannelGroupType(this.channelGroupTypeUID, this.advanced, this.label,
-                this.description, toChannelDefinitions(this.channelTypeReferences, channelTypes));
+                this.description, toChannelDefinitions(this.channelTypeReferences));
 
         return channelGroupType;
     }

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelXmlResult.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelXmlResult.java
@@ -8,13 +8,19 @@
 package org.eclipse.smarthome.core.thing.xml.internal;
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 import org.eclipse.smarthome.config.xml.util.NodeValue;
+import org.eclipse.smarthome.core.thing.type.ChannelDefinition;
+import org.eclipse.smarthome.core.thing.type.ChannelTypeUID;
+
+import com.thoughtworks.xstream.converters.ConversionException;
 
 /**
  * The {@link ChannelXmlResult} is an intermediate XML conversion result object.
- * 
+ *
  * @author Chris Jackson - Initial Contribution
  */
 public class ChannelXmlResult {
@@ -27,7 +33,7 @@ public class ChannelXmlResult {
 
     /**
      * Constructs a new {@link ChannelXmlResult}
-     * 
+     *
      * @param id the channel id
      * @param typeId the channel type id
      * @param label the channel label
@@ -44,7 +50,7 @@ public class ChannelXmlResult {
 
     /**
      * Retrieves the ID for this channel
-     * 
+     *
      * @return channel id
      */
     public String getId() {
@@ -53,7 +59,7 @@ public class ChannelXmlResult {
 
     /**
      * Retrieves the type ID for this channel
-     * 
+     *
      * @return type ID
      */
     public String getTypeId() {
@@ -62,7 +68,7 @@ public class ChannelXmlResult {
 
     /**
      * Retrieves the properties for this channel
-     * 
+     *
      * @return properties list (not null)
      */
     public List<NodeValue> getProperties() {
@@ -74,7 +80,7 @@ public class ChannelXmlResult {
 
     /**
      * Get the label for this channel
-     * 
+     *
      * @return the channel label. Can be null
      */
     public String getLabel() {
@@ -83,7 +89,7 @@ public class ChannelXmlResult {
 
     /**
      * Get the description for this channel
-     * 
+     *
      * @return the channel description. Can be null
      */
     public String getDescription() {
@@ -94,4 +100,31 @@ public class ChannelXmlResult {
     public String toString() {
         return "ChannelTypeXmlResult [id=" + id + ", typeId=" + typeId + ", properties=" + properties + "]";
     }
+
+    protected ChannelDefinition toChannelDefinition(String bindingId) throws ConversionException {
+        String id = getId();
+        String typeId = getTypeId();
+
+        String typeUID = getTypeUID(bindingId, typeId);
+
+        // Convert the channel properties into a map
+        Map<String, String> propertiesMap = new HashMap<>();
+        for (NodeValue property : getProperties()) {
+            propertiesMap.put(property.getAttributes().get("name"), (String) property.getValue());
+        }
+
+        ChannelDefinition channelDefinition = new ChannelDefinition(id, new ChannelTypeUID(typeUID), propertiesMap,
+                getLabel(), getDescription());
+
+        return channelDefinition;
+    }
+
+    private String getTypeUID(String bindingId, String typeId) {
+        if (typeId.startsWith(XmlHelper.SYSTEM_NAMESPACE_PREFIX)) {
+            return XmlHelper.getSystemUID(typeId);
+        } else {
+            return String.format("%s:%s", bindingId, typeId);
+        }
+    }
+
 }

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeXmlProvider.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeXmlProvider.java
@@ -67,7 +67,7 @@ public class ThingTypeXmlProvider implements XmlDocumentProvider<List<?>> {
 
     public ThingTypeXmlProvider(Bundle bundle, XmlConfigDescriptionProvider configDescriptionProvider,
             XmlThingTypeProvider thingTypeProvider, XmlChannelTypeProvider channelTypeProvider)
-                    throws IllegalArgumentException {
+            throws IllegalArgumentException {
 
         if (bundle == null) {
             throw new IllegalArgumentException("The Bundle must not be null!");
@@ -137,7 +137,7 @@ public class ThingTypeXmlProvider implements XmlDocumentProvider<List<?>> {
 
         // create channel group types
         for (ChannelGroupTypeXmlResult type : this.channelGroupTypeRefs) {
-            this.channelTypeProvider.addChannelGroupType(this.bundle, type.toChannelGroupType(channelTypes));
+            this.channelTypeProvider.addChannelGroupType(this.bundle, type.toChannelGroupType());
         }
 
         // create thing and bridge types

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeXmlResult.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeXmlResult.java
@@ -20,7 +20,6 @@ import org.eclipse.smarthome.core.thing.ThingTypeUID;
 import org.eclipse.smarthome.core.thing.type.ChannelDefinition;
 import org.eclipse.smarthome.core.thing.type.ChannelGroupDefinition;
 import org.eclipse.smarthome.core.thing.type.ChannelGroupTypeUID;
-import org.eclipse.smarthome.core.thing.type.ChannelTypeUID;
 import org.eclipse.smarthome.core.thing.type.ThingType;
 
 import com.thoughtworks.xstream.converters.ConversionException;
@@ -79,27 +78,8 @@ public class ThingTypeXmlResult {
             channelTypeDefinitions = new ArrayList<>(channelTypeReferences.size());
 
             for (ChannelXmlResult channelTypeReference : channelTypeReferences) {
-                String id = channelTypeReference.getId();
-                String typeId = channelTypeReference.getTypeId();
-
-                String typeUID = String.format("%s:%s", this.thingTypeUID.getBindingId(), typeId);
-
-                int systemPrefixIdx = typeId.indexOf(XmlHelper.SYSTEM_NAMESPACE_PREFIX);
-                if (systemPrefixIdx != -1) {
-                    typeUID = XmlHelper.getSystemUID(typeId);
-                }
-
-                // Convert the channel properties into a map
-                Map<String, String> propertiesMap = new HashMap<>();
-                for (NodeValue property : channelTypeReference.getProperties()) {
-                    propertiesMap.put(property.getAttributes().get("name"), (String) property.getValue());
-                }
-
-                ChannelDefinition channelDefinition = new ChannelDefinition(id, new ChannelTypeUID(typeUID),
-                        propertiesMap, channelTypeReference.getLabel(), channelTypeReference.getDescription());
-                channelTypeDefinitions.add(channelDefinition);
+                channelTypeDefinitions.add(channelTypeReference.toChannelDefinition(this.thingTypeUID.getBindingId()));
             }
-
         }
 
         return channelTypeDefinitions;

--- a/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/XmlHelper.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/XmlHelper.java
@@ -16,7 +16,7 @@ package org.eclipse.smarthome.core.thing.xml.internal;
 public class XmlHelper {
 
     public static final String SYSTEM_NAMESPACE_PREFIX = "system.";
-    public static final String SYSTEM_NAMESPACE = "system";
+    private static final String SYSTEM_NAMESPACE = "system";
 
     /**
      * Returns a UID in the format of {1}:{2}, where {1} is {@link #SYSTEM_NAMESPACE} and {2} is the
@@ -26,12 +26,9 @@ public class XmlHelper {
      * @return system uid (e.g. "system:test")
      */
     public static String getSystemUID(String typeId) {
-        int prefixIdx = typeId.indexOf(SYSTEM_NAMESPACE_PREFIX);
-
-        if (prefixIdx != -1) {
-            typeId = typeId.substring(prefixIdx + SYSTEM_NAMESPACE_PREFIX.length());
+        if (typeId.startsWith(SYSTEM_NAMESPACE_PREFIX)) {
+            typeId = typeId.substring(SYSTEM_NAMESPACE_PREFIX.length());
         }
-
         return String.format("%s:%s", SYSTEM_NAMESPACE, typeId);
     }
 }

